documentation/man/man3/Personage.3
.TH "Personage" 3 "Samedi 30 Juillet 2022" "Jeu POA" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Personage
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <Personage\&.hpp>\fP
.PP
Est dérivée de \fBMovableGameObject\fP\&.
.PP
Dérivée par \fBDruid\fP, \fBReligious\fP, \fBSoldier\fP, et \fBWorker\fP\&.
.SS "Fonctions membres publiques"

.in +1c
.ti -1c
.RI "\fBPersonage\fP (bool isPlayer=false)"
.br
.ti -1c
.RI "\fBPersonage\fP (\fBPersonage\fP const &p)=delete"
.br
.ti -1c
.RI "virtual \fB~Personage\fP ()"
.br
.ti -1c
.RI "bool \fBaddItem\fP (\fBItem\fP *item, \fBgame::inventory_items_types\fP type, sf::Vector2f const &coordinates)"
.br
.ti -1c
.RI "bool \fBmoveItem\fP (\fBItem\fP *item, \fBgame::inventory_items_types\fP type, \fBInventory\fP *to_inventory, sf::Vector2f const &coordinates)"
.br
.ti -1c
.RI "bool \fBremoveItem\fP (\fBItem\fP *item, \fBgame::inventory_items_types\fP type)"
.br
.ti -1c
.RI "bool \fBhaveThisItem\fP (\fBgame::ItemsCategories\fP const &categorie) const"
.br
.ti -1c
.RI "\fBgame::inventory_items_types\fP const & \fBgetInventoryItemType\fP (\fBItem\fP const *item) const"
.br
.ti -1c
.RI "\fBItem\fP const * \fBgetFirstItemMatch\fP (\fBgame::ItemsCategories\fP const &categorie) const"
.br
.ti -1c
.RI "\fBItem\fP * \fBgetFirstItemMatchNonConst\fP (\fBgame::ItemsCategories\fP const &categorie)"
.br
.ti -1c
.RI "const int & \fBgetHealthLevel\fP () const"
.br
.ti -1c
.RI "const int & \fBgetExperienceLevel\fP () const"
.br
.ti -1c
.RI "bool \fBisDied\fP () const"
.br
.ti -1c
.RI "virtual void \fBshow\fP (sf::RenderTarget &target) override"
.br
.RI "Pour dessiner l'élement dans la fenêtre\&. "
.ti -1c
.RI "void \fBupdate\fP () override"
.br
.ti -1c
.RI "void \fBupdateMousePosWindow\fP (sf::Vector2i mousePosWindow)"
.br
.ti -1c
.RI "void \fBonCollisionEnter\fP (\fBCollision\fP const &collision) const override"
.br
.RI "émit quand il entre en contacte avec un autre élément "
.ti -1c
.RI "virtual void \fBmove\fP (\fBMyVector\fP &speed) override"
.br
.ti -1c
.RI "void \fBrenderBagInventory\fP (sf::RenderTarget &target)"
.br
.ti -1c
.RI "virtual void \fBsimpleAttack\fP (\fBPersonage\fP &target)=0"
.br
.ti -1c
.RI "virtual void \fBspecialAttack\fP (\fBPersonage\fP &target)=0"
.br
.ti -1c
.RI "virtual \fBgame::ItemActionResult\fP \fBreceiveItemAction\fP (int value, \fBgame::ItemsCategories\fP categorie)"
.br
.ti -1c
.RI "\fBgame::AnimationSide\fP const & \fBgetLastAnimationSide\fP ()"
.br
.RI "Get the Last \fBAnimation\fP Side\&. "
.ti -1c
.RI "void \fBsetLastAnimationSide\fP (\fBgame::AnimationSide\fP side)"
.br
.in -1c
.SS "Fonctions membres publiques statiques"

.in +1c
.ti -1c
.RI "static PersonageMap \fBgetPersonages\fP ()"
.br
.in -1c
.SS "Fonctions membres protégées"

.in +1c
.ti -1c
.RI "void \fBinit\fP ()"
.br
.RI "Pour initialiser les valeur par défaut du \fBGameObject\fP\&. "
.ti -1c
.RI "void \fBinitAnimations\fP () override"
.br
.RI "Pour ajouter les animation pour un personnage\&. "
.ti -1c
.RI "void \fBinitEXPBar\fP ()"
.br
.ti -1c
.RI "void \fBinitHPBar\fP ()"
.br
.ti -1c
.RI "void \fBinitBag\fP ()"
.br
.ti -1c
.RI "sf::Vector2f \fBcalculateProgressBarsPosition\fP ()"
.br
.ti -1c
.RI "void \fBincreaseHealth\fP (int value)"
.br
.RI "Augmenter la santé "
.ti -1c
.RI "void \fBincreaseExp\fP (int value)"
.br
.RI "Augmenter l'expérience\&. "
.ti -1c
.RI "void \fBdecreaseHealth\fP (int value)"
.br
.RI "Diminuer la santé "
.ti -1c
.RI "void \fBdecreaseExp\fP (int value)"
.br
.RI "Diminuer l'expériance\&. "
.ti -1c
.RI "bool \fBisLastAnimationSideRight\fP ()"
.br
.ti -1c
.RI "void \fBplayCorrectSideHurtAnimation\fP ()"
.br
.RI "Pour l'animation HURT en tenant compte du coté vers lequel est dirigé le personage\&. "
.ti -1c
.RI "virtual void \fBupdatePosition\fP (float posX, float posY) override"
.br
.RI "Permet de mettre à jour la position actuelle\&. Cette position sera déterminé par l'évenement reçu au préalable\&. Aussi il s'agit dans ce cas d'une accélération uniforme\&. "
.ti -1c
.RI "virtual unsigned \fBreceiveHealthDamage\fP (int value)"
.br
.RI "pour recevoir les dégats de santé Retourne les dégats réelemnt concidéré (au cas ou la personne utilise un bouclié ou\&.\&.\&.) "
.ti -1c
.RI "virtual unsigned \fBreceiveExpDamage\fP (int value)"
.br
.RI "pour recevoir les dégats d'exp Retourne les dégats réelemnt concidéré (au cas ou la personne utilise un bouclié ou\&.\&.\&.) "
.ti -1c
.RI "virtual unsigned \fBreceiveHealthCare\fP (int value)"
.br
.RI "pour recevoir les soins santé Retourne les dégats réelemnt concidéré (au cas ou la personne utilise un bouclié ou\&.\&.\&.) "
.ti -1c
.RI "virtual unsigned \fBreceiveExpCare\fP (int value)"
.br
.RI "pour recevoir les soins Exp Retourne les dégats réelemnt concidéré (au cas ou la personne utilise un bouclié ou\&.\&.\&.) "
.ti -1c
.RI "virtual void \fBreceiveShieldboost\fP (int value)"
.br
.RI "Pour activer la protection du bouclier\&. Il faut noter que son activation est temporaire et protège la santé en diminuant l'ataque\&. "
.ti -1c
.RI "void \fBupdateEXPBar\fP ()"
.br
.ti -1c
.RI "void \fBupdateHPBar\fP ()"
.br
.ti -1c
.RI "void \fBrenderEXPBar\fP (sf::RenderTarget &target)"
.br
.ti -1c
.RI "void \fBrenderHPBar\fP (sf::RenderTarget &target)"
.br
.in -1c
.SS "Attributs protégés"

.in +1c
.ti -1c
.RI "\fBgame::AnimationSide\fP \fBlastAnimationSide\fP"
.br
.RI "Permet de controler de quelque coté diriger une animation (gauche ou droite) "
.ti -1c
.RI "\fBBag\fP * \fBbag\fP"
.br
.ti -1c
.RI "int \fBm_healthLevel\fP { 100 }"
.br
.ti -1c
.RI "int \fBm_experienceLevel\fP { 50 }"
.br
.RI "La quantité de ça spécialité\&. NOTE -\&. "
.ti -1c
.RI "int \fBm_shieldBoost\fP { 0 }"
.br
.RI "représente la valeur de protection offerte par le bouclié\&. NOTE - Meme si la valeur n'est pas nulle, tant que le bouclié n'est pas activé, cette dernière n'est pas concidérée "
.ti -1c
.RI "std::string \fBm_specialtyName\fP"
.br
.ti -1c
.RI "int \fBm_simpleAttackValue\fP { 10 }"
.br
.ti -1c
.RI "int \fBm_specialAttackValue\fP { 25 }"
.br
.ti -1c
.RI "float \fBm_probabilitySuccessAttack\fP { \&.5f }"
.br
.ti -1c
.RI "float \fBm_probabilityDodge\fP { \&.1 }"
.br
.ti -1c
.RI "int \fBm_attackVariation\fP { 5 }"
.br
.ti -1c
.RI "sf::Sprite \fBm_healthBarSprite\fP"
.br
.ti -1c
.RI "sf::Sprite \fBm_specialityBarSprite\fP"
.br
.ti -1c
.RI "sf::Sprite \fBm_suit\fP"
.br
.RI "Représente la tenue du personage\&. REVIEW\&. "
.ti -1c
.RI "const bool \fBm_isPlayer\fP"
.br
.ti -1c
.RI "sf::RenderTarget * \fBm_window\fP { 0 }"
.br
.ti -1c
.RI "\fBgui::ProgressBar\fP * \fBexpBar\fP"
.br
.ti -1c
.RI "\fBgui::ProgressBar\fP * \fBhpBar\fP"
.br
.in -1c
.SS "Attributs protégés statiques"

.in +1c
.ti -1c
.RI "static PersonageMap \fBpersonages\fP"
.br
.in -1c
.SS "Membres hérités additionnels"
.SH "Description détaillée"
.PP 
Définition à la ligne \fB26\fP du fichier \fBPersonage\&.hpp\fP\&.
.SH "Documentation des constructeurs et destructeur"
.PP 
.SS "Personage::Personage (bool isPlayer = \fCfalse\fP)\fC [explicit]\fP"

.PP
Définition à la ligne \fB146\fP du fichier \fBPersonage\&.cpp\fP\&.
.SS "Personage::Personage (\fBPersonage\fP const & p)\fC [explicit]\fP, \fC [delete]\fP"

.SS "Personage::~Personage ()\fC [virtual]\fP"

.PP
Définition à la ligne \fB173\fP du fichier \fBPersonage\&.cpp\fP\&.
.SH "Documentation des fonctions membres"
.PP 
.SS "bool Personage::addItem (\fBItem\fP * item, \fBgame::inventory_items_types\fP type, sf::Vector2f const & coordinates)"

.PP
Définition à la ligne \fB189\fP du fichier \fBPersonage\&.cpp\fP\&.
.SS "sf::Vector2f Personage::calculateProgressBarsPosition ()\fC [protected]\fP"

.PP
Définition à la ligne \fB234\fP du fichier \fBPersonage\&.cpp\fP\&.
.SS "void Personage::decreaseExp (int value)\fC [protected]\fP"

.PP
Diminuer l'expériance\&. 
.PP
\fBParamètres\fP
.RS 4
\fIvalue\fP 
.RE
.PP

.PP
Définition à la ligne \fB287\fP du fichier \fBPersonage\&.cpp\fP\&.
.SS "void Personage::decreaseHealth (int value)\fC [protected]\fP"

.PP
Diminuer la santé 
.PP
\fBParamètres\fP
.RS 4
\fIvalue\fP 
.RE
.PP

.PP
Définition à la ligne \fB260\fP du fichier \fBPersonage\&.cpp\fP\&.
.SS "const int & Personage::getExperienceLevel () const"

.PP
Définition à la ligne \fB229\fP du fichier \fBPersonage\&.cpp\fP\&.
.SS "\fBItem\fP const  * Personage::getFirstItemMatch (\fBgame::ItemsCategories\fP const & categorie) const"

.PP
Définition à la ligne \fB214\fP du fichier \fBPersonage\&.cpp\fP\&.
.SS "\fBItem\fP * Personage::getFirstItemMatchNonConst (\fBgame::ItemsCategories\fP const & categorie)"

.PP
Définition à la ligne \fB219\fP du fichier \fBPersonage\&.cpp\fP\&.
.SS "const int & Personage::getHealthLevel () const"

.PP
Définition à la ligne \fB224\fP du fichier \fBPersonage\&.cpp\fP\&.
.SS "\fBgame::inventory_items_types\fP const  & Personage::getInventoryItemType (\fBItem\fP const * item) const"

.PP
Définition à la ligne \fB209\fP du fichier \fBPersonage\&.cpp\fP\&.
.SS "\fBgame::AnimationSide\fP const  & Personage::getLastAnimationSide ()"

.PP
Get the Last \fBAnimation\fP Side\&. 
.PP
\fBRenvoie\fP
.RS 4
\fBgame::AnimationSide\fP 
.RE
.PP

.PP
Définition à la ligne \fB392\fP du fichier \fBPersonage\&.cpp\fP\&.
.SS "Personage::PersonageMap Personage::getPersonages ()\fC [static]\fP"

.PP
Définition à la ligne \fB402\fP du fichier \fBPersonage\&.cpp\fP\&.
.SS "bool Personage::haveThisItem (\fBgame::ItemsCategories\fP const & categorie) const"

.PP
Définition à la ligne \fB204\fP du fichier \fBPersonage\&.cpp\fP\&.
.SS "void Personage::increaseExp (int value)\fC [protected]\fP"

.PP
Augmenter l'expérience\&. 
.PP
\fBParamètres\fP
.RS 4
\fIvalue\fP 
.RE
.PP

.PP
Définition à la ligne \fB250\fP du fichier \fBPersonage\&.cpp\fP\&.
.SS "void Personage::increaseHealth (int value)\fC [protected]\fP"

.PP
Augmenter la santé 
.PP
\fBParamètres\fP
.RS 4
\fIvalue\fP 
.RE
.PP

.PP
Définition à la ligne \fB240\fP du fichier \fBPersonage\&.cpp\fP\&.
.SS "void Personage::init ()\fC [protected]\fP"

.PP
Pour initialiser les valeur par défaut du \fBGameObject\fP\&. 
.PP
Définition à la ligne \fB36\fP du fichier \fBPersonage\&.cpp\fP\&.
.SS "void Personage::initAnimations ()\fC [override]\fP, \fC [protected]\fP, \fC [virtual]\fP"

.PP
Pour ajouter les animation pour un personnage\&. 
.PP
Implémente \fBMovableGameObject\fP\&.
.PP
Définition à la ligne \fB88\fP du fichier \fBPersonage\&.cpp\fP\&.
.SS "void Personage::initBag ()\fC [protected]\fP"

.PP
Définition à la ligne \fB122\fP du fichier \fBPersonage\&.cpp\fP\&.
.SS "void Personage::initEXPBar ()\fC [protected]\fP"

.PP
Définition à la ligne \fB110\fP du fichier \fBPersonage\&.cpp\fP\&.
.SS "void Personage::initHPBar ()\fC [protected]\fP"

.PP
Définition à la ligne \fB116\fP du fichier \fBPersonage\&.cpp\fP\&.
.SS "bool Personage::isDied () const"

.PP
Définition à la ligne \fB314\fP du fichier \fBPersonage\&.cpp\fP\&.
.SS "bool Personage::isLastAnimationSideRight ()\fC [protected]\fP"

.PP
Définition à la ligne \fB297\fP du fichier \fBPersonage\&.cpp\fP\&.
.SS "void Personage::move (\fBMyVector\fP & speed)\fC [override]\fP, \fC [virtual]\fP"

.PP
Implémente \fBMovableGameObject\fP\&.
.PP
Définition à la ligne \fB382\fP du fichier \fBPersonage\&.cpp\fP\&.

src/include/Personage.cpp
#include "include/Personage.hpp"
#include "include/BoxCollider.hpp"
#include "include/BoxingGloves.hpp"
#include "include/GameMaster.hpp"
#include "include/Money.hpp"
#include "include/MyVector.hpp"
#include "include/Rigidbody.hpp"
#include "include/Shield.hpp"
#include "include/Sword.hpp"
#include "include/TeleportKey.hpp"
#include "include/Vial.hpp"
#include "include/consts.hpp"
#include <iostream>
#include <string>

class Personage;
class BoxCollider;
class BoxingGloves;
class GameMaster;
class Money;
struct MyVector;
class Rigidbody;
class Shield;
class Sword;
class TeleportKey;
class Vial;

using namespace std;

// Initialisation à vide de la liste de personages
Personage::PersonageMap Personage::personages = [] {
	PersonageMap ret;
	return ret;
}();

void Personage::init()
{
	// m_width = 32;
	// m_height = 50;

	// m_gameObjectName = "";
	this->setGameObjectName("Personage");

	// m_texture.setSmooth(true);

	// m_body.setScale(0.3f, 0.3f);
	// m_body.setColor(sf::Color::White);

	// On définit son centre de "gravité"
	// m_body.setOrigin(m_body.getLocalBounds().width / 2, m_body.getLocalBounds().height / 2);
	// m_body.setOrigin(PERSONNAGE_WIDTH / 2, PERSONNAGE_WIDTH / 2);

	// m_collider = new PolygonCollider((*this), { sf::Vector2f(18.f - GameMaster::getSPRITE_BOX_CENTER().x, 5.f - GameMaster::getSPRITE_BOX_CENTER().y), sf::Vector2f(22.f - GameMaster::getSPRITE_BOX_CENTER().x, 12.f - GameMaster::getSPRITE_BOX_CENTER().y), sf::Vector2f(22.f - GameMaster::getSPRITE_BOX_CENTER().x, 20.f - GameMaster::getSPRITE_BOX_CENTER().y), sf::Vector2f(18.f - GameMaster::getSPRITE_BOX_CENTER().x, 27.f - GameMaster::getSPRITE_BOX_CENTER().y), sf::Vector2f(11.f - GameMaster::getSPRITE_BOX_CENTER().x, 22.f - GameMaster::getSPRITE_BOX_CENTER().y), sf::Vector2f(11.f - GameMaster::getSPRITE_BOX_CENTER().x, 12.f - GameMaster::getSPRITE_BOX_CENTER().y) });
	// m_collider = new PolygonCollider((*this), { sf::Vector2f(18.f - m_body.getOrigin().x, 5.f - m_body.getOrigin().y), sf::Vector2f(22.f - m_body.getOrigin().x, 12.f - m_body.getOrigin().y), sf::Vector2f(22.f - m_body.getOrigin().x, 20.f - m_body.getOrigin().y), sf::Vector2f(18.f - m_body.getOrigin().x, 27.f - m_body.getOrigin().y), sf::Vector2f(11.f - m_body.getOrigin().x, 22.f - m_body.getOrigin().y), sf::Vector2f(11.f - m_body.getOrigin().x, 12.f - m_body.getOrigin().y) });

	m_rigidbody = new Rigidbody((*this));
	m_collider = new BoxCollider((*this), 45.f, 95.f, sf::Vector2f(75.f - 45.f / 2.f, 115.f - 95.f));

	// int w(32);
	// int h(32);
	// m_body.setTextureRect(sf::IntRect(0, 0, w, h)); // REVIEW - Un personnage par défaut plus stylé

	// m_body.setScale((float)PERSONNAGE_WIDTH / (float)w, (float)PERSONNAGE_WIDTH / (float)h);

	// On définit l'origine du sprite (c'est a partir de cette dernière qu'on calcul la position du joueur.)
	m_body.setOrigin(75.f, 115.f);

	// Initialisation de la position en temps que joueurs
	// REVIEW - Cela se fera après l'ajout dans une salle
	// if (m_isPlayer)
	// {
	// 	cout << "Placement du joueur sur le lieu de départ\n";

	// 	m_position->setPosition(game::GAME_BLOCKS_WIDTH * 4, game::GAME_BLOCKS_WIDTH * 78); // REVIEW -
	// 	// m_position = Position(game::GAME_BLOCKS_WIDTH * 4, game::GAME_BLOCKS_WIDTH * 4);
	// 	m_body.setPosition(m_position->getX(), m_position->getY());
	// }
	// else
	// {
	// 	// TODO - Sera un énémi donc on devra avoir une classe spéciale pour lui
	// 	m_position->setPosition(game::GAME_BLOCKS_WIDTH * 20, game::GAME_BLOCKS_WIDTH * 70); // REVIEW -
	// 	m_body.setPosition(m_position->getX(), m_position->getY());
	// }

	// m_body.setScale(.3f, .3f);
}

void Personage::initAnimations()
{
	// Création du gestionnaire d'animation
	this->createAnimationComponent(m_texture);
	// Ajout des animations
	this->animationComponent->addAnimation(4, "RIGHT_IDLE", 10.f, 0, 0, 0, 0, 150, 150);
	this->animationComponent->addAnimation(3, "RIGHT_WALK", 10.f, 0, 0, 3, 0, 150, 150);				//
	this->animationComponent->addAnimation(2, "RIGHT_ATTACK_BOXING_GLOVES", 5.f, 0, 2, 8, 2, 150, 150); // Une fois
	this->animationComponent->addAnimation(2, "RIGHT_ATTACK_SWORD", 7.f, 0, 4, 4, 4, 150, 150);			// Une fois
	this->animationComponent->addAnimation(2, "RIGHT_DEFEND_SHIELD", 6.f, 0, 6, 3, 6, 150, 150, true);	// Une fois et reste maintenu
	this->animationComponent->addAnimation(1, "RIGHT_HURT", 5.f, 0, 8, 1, 8, 150, 150);					// Une fois
	this->animationComponent->addAnimation(0, "RIGHT_DIE", 10.f, 0, 10, 7, 10, 150, 150, true);			// Une fois et reste maintenu

	this->animationComponent->addAnimation(4, "LEFT_IDLE", 10.f, 0, 1, 0, 1, 150, 150);
	this->animationComponent->addAnimation(3, "LEFT_WALK", 10.f, 0, 1, 3, 1, 150, 150);				   //
	this->animationComponent->addAnimation(2, "LEFT_ATTACK_BOXING_GLOVES", 5.f, 0, 3, 8, 3, 150, 150); // Une fois
	this->animationComponent->addAnimation(2, "LEFT_ATTACK_SWORD", 7.f, 0, 5, 4, 5, 150, 150);		   // Une fois
	this->animationComponent->addAnimation(2, "LEFT_DEFEND_SHIELD", 6.f, 0, 7, 3, 7, 150, 150, true);  // Une fois et reste maintenu
	this->animationComponent->addAnimation(1, "LEFT_HURT", 5.f, 0, 9, 1, 9, 150, 150);				   // Une fois
	this->animationComponent->addAnimation(0, "LEFT_DIE", 10.f, 0, 11, 7, 11, 150, 150, true);		   // Une fois et reste maintenu
}

void Personage::initEXPBar()
{
	this->expBar = new gui::ProgressBar(
		this->calculateProgressBarsPosition().x - 10.4f * 5.5f, this->calculateProgressBarsPosition().y + 5.6f * 5, 10.4f, 1.9f, sf::Color::Blue, 220, GameMaster::STATE_DATA()->graphicsSettings->resolution, &GameMaster::STATE_DATA()->defaultFont);
}

void Personage::initHPBar()
{
	this->hpBar = new gui::ProgressBar(
		this->calculateProgressBarsPosition().x - 10.4f * 5.5f, this->calculateProgressBarsPosition().y + 8.3f, 10.4f, 2.8f, sf::Color::Red, 180, GameMaster::STATE_DATA()->graphicsSettings->resolution, &GameMaster::STATE_DATA()->defaultFont);
}

void Personage::initBag()
{
	this->bag = new Bag();

	// Satrt Vials
	this->bag->addItem(new Vial(game::ItemsCategories::VIAL_HEALTH), game::inventory_items_types::VIAL, sf::Vector2f());
	this->bag->addItem(new Vial(game::ItemsCategories::VIAL_EXP), game::inventory_items_types::VIAL, sf::Vector2f());
	this->bag->addItem(new Vial(game::ItemsCategories::VIAL_ATTACK_HEALTH), game::inventory_items_types::VIAL, sf::Vector2f());
	this->bag->addItem(new Vial(game::ItemsCategories::VIAL_ATTACK_EXP), game::inventory_items_types::VIAL, sf::Vector2f());

	// Start Armoies
	this->bag->addItem(new Shield(), game::inventory_items_types::ARMORY, sf::Vector2f());
	this->bag->addItem(new BoxingGloves(), game::inventory_items_types::ARMORY, sf::Vector2f());
	this->bag->addItem(new Sword(), game::inventory_items_types::ARMORY, sf::Vector2f());

	// Start teleportkey
	this->bag->addItem(new TeleportKey(), game::inventory_items_types::TELEPORTKEY, sf::Vector2f());

	// Start money "3"
	this->bag->addItem(new Money(), game::inventory_items_types::MONEY, sf::Vector2f());
	this->bag->addItem(new Money(), game::inventory_items_types::MONEY, sf::Vector2f());
	this->bag->addItem(new Money(), game::inventory_items_types::MONEY, sf::Vector2f());
}

Personage::Personage(bool isPlayer) :
	MovableGameObject("content/personage/personage.png"), // REVIEW -
	m_isPlayer(isPlayer)
{
	this->init();
	this->initAnimations();
	this->initEXPBar();
	this->initHPBar();
	this->initBag();

	Personage::personages.insert({ std::to_string(this->getGameObjectId()), this });
}

// Personage::Personage(std::string_view const& imageSpritePath, bool isPlayer) :
// 	GameObject(imageSpritePath),
// 	m_isPlayer(isPlayer)
// {
// 	init();
// }

// Personage::Personage(float posX, float posY, bool isPlayer) :
// 	GameObject(posX, posY),
// 	m_isPlayer(isPlayer)
// {
// 	init();
// }

Personage::~Personage()
{
	// On vide le sac vers la salle actuelle
	// TODO -

	// delete m_collider; // Déjà fait au niveau du GameObject.
	m_window = nullptr;
	delete this->hpBar;
	delete this->expBar;
	delete this->bag;

	// On supprime cet personage de la liste static des personnages
	Personage::personages.erase(std::to_string(this->getGameObjectId()));
}

//
bool Personage::addItem(Item* item, game::inventory_items_types type, sf::Vector2f const& coordinates)
{
	return this->bag->addItem(item, type, coordinates);
}

bool Personage::moveItem(Item* item, game::inventory_items_types type, Inventory* to_inventory, sf::Vector2f const& coordinates)
{
	return this->bag->moveItem(item, type, to_inventory, coordinates);
}

bool Personage::removeItem(Item* item, game::inventory_items_types type)
{
	return this->bag->removeItem(item, type);
}

bool Personage::haveThisItem(game::ItemsCategories const& categorie) const
{
	return this->bag->haveThisItem(categorie);
}

game::inventory_items_types const& Personage::getInventoryItemType(Item const* item) const
{
	return this->bag->getInventoryItemType(item);
}

Item const* Personage::getFirstItemMatch(game::ItemsCategories const& categorie) const
{
	return this->bag->getFirstItemMatch(categorie);
}

Item* Personage::getFirstItemMatchNonConst(game::ItemsCategories const& categorie)
{
	return this->bag->getFirstItemMatchNonConst(categorie);
}

const int& Personage::getHealthLevel() const
{
	return m_healthLevel;
}

const int& Personage::getExperienceLevel() const
{
	return m_experienceLevel;
}

sf::Vector2f Personage::calculateProgressBarsPosition()
{
	return sf::Vector2f(this->getPosition()->getPosition().x, this->getPosition()->getPosition().y - 150.f);
}

//* --------------------------- Gestion augmentation et diminution de la vie
void Personage::increaseHealth(int value)
{
	m_healthLevel += value;

	if (m_healthLevel > game::PERSONNAGE_MAX_HEALTH)
	{
		m_healthLevel = game::PERSONNAGE_MAX_HEALTH;
	}
}

void Personage::increaseExp(int value)
{
	m_experienceLevel += value;

	if (m_experienceLevel > game::PERSONNAGE_MAX_EXP)
	{
		m_experienceLevel = game::PERSONNAGE_MAX_EXP;
	}
}

void Personage::decreaseHealth(int value)
{
	if (m_healthLevel == 0)
	{
		return;
	}

	m_healthLevel -= value;

	// le = est nécessaire pour faire l'animation même quand le m_healthLevel arrive directement à z&ro
	if (m_healthLevel <= 0)
	{
		m_healthLevel = 0;
		// TODO - Méthode die et tous ce qui va avec
		std::string animationName = "DIE";
		if (this->isLastAnimationSideRight())
		{
			animationName = "RIGHT_" + animationName;
		}
		else
		{
			animationName = "LEFT_" + animationName;
		}
		this->getAnimationComponent()->play(animationName, 0.016417f);

		// On drop tout les items
		this->bag->dropAll(this->getPosition()->getPosition().toVector2f());
	}
}

void Personage::decreaseExp(int value)
{
	m_experienceLevel -= value;

	if (m_experienceLevel < 0)
	{
		m_experienceLevel = 0;
	}
}

bool Personage::isLastAnimationSideRight()
{
	return this->getLastAnimationSide() == game::AnimationSide::RIGHT;
}

void Personage::playCorrectSideHurtAnimation()
{
	if (this->isLastAnimationSideRight())
	{
		this->getAnimationComponent()->play("RIGHT_HURT", 0.016417f);
	}
	else
	{
		this->getAnimationComponent()->play("LEFT_HURT", 0.016417f);
	}
}

bool Personage::isDied() const
{
	return m_healthLevel == 0; // REVIEW -
}

void Personage::show(sf::RenderTarget& target)
{
	m_window = &target; // REVIEW -
	// GameObject::show(window);
	target.draw(m_body);

	this->renderEXPBar(target);
	this->renderHPBar(target);
}

void Personage::update()
{
	this->updateEXPBar();
	this->updateHPBar();

	this->bag->getInventory()->getGui()->update();
}

void Personage::updateMousePosWindow(sf::Vector2i mousePosWindow)
{
	this->bag->getInventory()->getGui()->updateMousePosWindow(mousePosWindow);
}

// void Personage::setGameObjectName(string name)
// {
// 	m_gameObjectName = name;
// }

void Personage::onCollisionEnter(Collision const& collision) const
{
	cout << "COLLISION !!!!!\n";
	if (m_isPlayer)
	{
		cout << "Collision de " << m_gameObjectName << " avec "
			 << collision.getGameObject().getGameObjectName() << endl;
	}
}

void Personage::updatePosition(float posX, float posY)
{
	m_position->setPosition(posX, posY);
	// On replace le body
	m_body.setPosition(m_position->getPosition().toVector2f());

	// On déplace les progresses
	this->hpBar->setPosition(sf::Vector2f(this->calculateProgressBarsPosition().x - 10.4f * 5.5f, this->calculateProgressBarsPosition().y + 8.3f), GameMaster::STATE_DATA()->graphicsSettings->resolution);
	this->expBar->setPosition(sf::Vector2f(this->calculateProgressBarsPosition().x - 10.4f * 5.5f, this->calculateProgressBarsPosition().y + 5.6f * 5), GameMaster::STATE_DATA()->graphicsSettings->resolution);

	// On replace la vue si c'est le joueur
	if (m_isPlayer)
		if (m_window != 0)
		{
			// cout << "On replace la vue" << endl;
			sf::View player_view(m_window->getView());
			// player_view.setCenter(m_position.getX(), m_position.getY());
			player_view.setCenter(this->m_position->getPosition().toVector2f());
			m_window->setView(player_view);
		}

	cout << "Position du joueur " << m_gameObjectName << " : " << m_position << " => " << (*m_position) << endl;
}

void Personage::move(MyVector& speed)
{
	updatePosition(this->getPosition()->getX() + speed.x, this->getPosition()->getY() + speed.y);
}

void Personage::renderBagInventory(sf::RenderTarget& target)
{
	this->bag->renderInventory(target);
}

game::AnimationSide const& Personage::getLastAnimationSide()
{
	return this->lastAnimationSide;
}

void Personage::setLastAnimationSide(game::AnimationSide side)
{
	this->lastAnimationSide = side;
}

Personage::PersonageMap Personage::getPersonages()
{
	return Personage::personages;
}

//* ----------------------------  Gestion des effets d'item (interne externe)
game::ItemActionResult Personage::receiveItemAction(int value, game::ItemsCategories categorie)
{
	game::ItemActionResult res;
	res.healthDecrease = 0;
	res.healthIncrease = 0;
	res.experienceDecrease = 0;
	res.experienceincrease = 0;
	res.isDied = false;

	switch (categorie)
	{
		case game::ItemsCategories::DEFEND:
			// TODO -
			this->receiveShieldboost(value);
			break;
		case game::ItemsCategories::ATTACK_BODY_TO_BODY:
			res.healthDecrease = this->receiveHealthDamage(value);
			this->playCorrectSideHurtAnimation();
			break;
		case game::ItemsCategories::ATTACK_SEMI_DISTANCE:
			res.healthDecrease = this->receiveHealthDamage(value);
			this->playCorrectSideHurtAnimation();
			break;
		case game::ItemsCategories::ATTACK_DISTANCE:
			res.healthDecrease = this->receiveHealthDamage(value);
			this->playCorrectSideHurtAnimation();
			break;
		case game::ItemsCategories::VIAL_EXP:
			res.experienceincrease = this->receiveExpCare(value);
			// TODO - Une animation peut ètre
			break;
		case game::ItemsCategories::VIAL_HEALTH:
			res.healthIncrease = this->receiveHealthCare(value);
			// TODO -
			break;
		case game::ItemsCategories::VIAL_ATTACK_EXP:
			res.experienceDecrease = this->receiveExpDamage(value);
			this->playCorrectSideHurtAnimation();
			break;
		case game::ItemsCategories::VIAL_ATTACK_HEALTH:
			res.healthDecrease = this->receiveHealthDamage(value);
			this->playCorrectSideHurtAnimation();
			break;

		default:
			break;
	}

	res.isDied = this->isDied();
	return res;
}
// TODO - Utiliser tous les systèmesde complexification de combat
unsigned Personage::receiveHealthDamage(int value)
{
	this->decreaseHealth(value);

	return value;
}

unsigned Personage::receiveExpDamage(int value)
{
	this->decreaseExp(value);

	return value;
}

unsigned Personage::receiveHealthCare(int value)
{
	this->increaseHealth(value);

	return value;
}

unsigned Personage::receiveExpCare(int value)
{
	this->increaseExp(value);

	return value;
}

void Personage::receiveShieldboost(int value)
{
	// REVIEW -

	this->m_shieldBoost = value;
}

void Personage::updateEXPBar()
{
	this->expBar->update(this->m_experienceLevel, game::PERSONNAGE_MAX_EXP);
}

void Personage::updateHPBar()
{
	this->hpBar->update(this->m_healthLevel, game::PERSONNAGE_MAX_HEALTH);
}

void Personage::renderEXPBar(sf::RenderTarget& target)
{
	this->expBar->render(target);
}

void Personage::renderHPBar(sf::RenderTarget& target)
{
	this->hpBar->render(target);
}
